rql/parser.go

```diff
@@ -0,0 +1,200 @@
+			err := validateNumberType(filterItem)
+			if err != nil {
+				return err
+			}
+		case "bool":
+			err := validateBoolType(filterItem)
+			if err != nil {
+				return err
+			}
+		case "datetime":
+			err := validateDatetimeType(filterItem)
+			if err != nil {
+				return err
+			}
+		case "string":
+			err := validateStringType(filterItem)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("type '%s' is not recognized", allowedDataType)
+		}
+
+		if !isValidOperator(filterItem) {
+			return fmt.Errorf("value '%s' for key '%s' is valid string", filterItem.Operator, filterItem.Name)
+		}
+	}
+
+	err := validateGroupByKeys(q, val)
+	if err != nil {
+		return err
+	}
+	return validateSortKey(q, val)
+}
+
+func validateNumberType(filterItem Filter) error {
+	// check if the type is any of Golang numeric types
+	// if not, return error
+	switch filterItem.Value.(type) {
+	case uint8, uint16, uint32, uint64, int8, int16, int32, int64, float32, float64, int, uint:
+		return nil
+	default:
+		return fmt.Errorf("value %v for key '%s' is not int type", filterItem.Value, filterItem.Name)
+	}
+}
+
+func validateDatetimeType(filterItem Filter) error {
+	// cast the value to datetime
+	// if failed, return error
+	castedVal, ok := filterItem.Value.(string)
+	if !ok {
+		return fmt.Errorf("value %s for key '%s' is not a valid ISO datetime string", filterItem.Value, filterItem.Name)
+	}
+	_, err := time.Parse(time.RFC3339, castedVal)
+	if err != nil {
+		return fmt.Errorf("value %s for key '%s' is not a valid ISO datetime string", filterItem.Value, filterItem.Name)
+	}
+	return nil
+}
+
+func validateBoolType(filterItem Filter) error {
+	// cast the value to bool
+	// if failed, return error
+	_, ok := filterItem.Value.(bool)
+	if !ok {
+		return fmt.Errorf("value %v for key '%s' is not bool type", filterItem.Value, filterItem.Name)
+	}
+	return nil
+}
+
+func validateStringType(filterItem Filter) error {
+	// cast the value to string
+	// if failed, return error
+	_, ok := filterItem.Value.(string)
+	if !ok {
+		return fmt.Errorf("value %s for key '%s' is valid string type", filterItem.Value, filterItem.Name)
+	}
+	return nil
+}
+
+func searchKeyInsideStruct(keyName string, val reflect.Value) int {
+	for i := 0; i < val.NumField(); i++ {
+		if strings.ToLower(val.Type().Field(i).Name) == strings.ToLower(keyName) {
+			return i
+		}
+	}
+	return -1
+}
+
+// parse the tag schema which is of the format
+// type=int,min=10,max=200
+// to extract type else fallback to string
+func getDataTypeOfField(tagString string) string {
+	res := "string"
+	splitted := strings.Split(tagString, ",")
+	for _, item := range splitted {
+		kvSplitted := strings.Split(item, "=")
+		if len(kvSplitted) == 2 {
+			if kvSplitted[0] == "type" {
+				return kvSplitted[1]
+			}
+		}
+	}
+	//fallback to string if type not found in tag value
+	return res
+}
+
+func isValidOperator(filterItem Filter) bool {
+	switch filterItem.dataType {
+	case "number":
```

Comment:
Can we have this as constants instead since we are using them in multiple places. Would also help in reading code where these dataType constants would be listed together.